File: app/src/main/java/lt/vilnius/tvarkau/fragments/SingleProblemMapFragment.java
package lt.vilnius.tvarkau.fragments;

import android.os.Bundle;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.Marker;

import org.parceler.Parcels;

import lt.vilnius.tvarkau.entity.Problem;

public class SingleProblemMapFragment extends BaseMapFragment implements OnMapReadyCallback,
        GoogleMap.OnInfoWindowClickListener,
        GoogleMap.OnInfoWindowCloseListener {

    Problem problem;

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        getMapAsync(this);
    }

    public static SingleProblemMapFragment getInstance(Problem problem) {
        SingleProblemMapFragment singleProblemMapFragment = new SingleProblemMapFragment();
        Bundle arguments = new Bundle();
        arguments.putParcelable(null, Parcels.wrap(problem));
        singleProblemMapFragment.setArguments(arguments);
        return singleProblemMapFragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            problem = Parcels.unwrap(getArguments().getParcelable(null));
        }
    }

    @Override
    protected void initMapData() {
        addSingleProblemMarker();
    }

    private void addSingleProblemMarker() {
        if (problem != null) {
            placeAndShowMarker(problem);
            setMapTitle(problem.getAddress());
        }
    }

    @Override
    public void onInfoWindowClick(Marker marker) {
        getActivity().onBackPressed();
    }

    @Override
    public void onInfoWindowClose(Marker marker) {
        Problem problem = getProblemByMarker(marker);

        getActivity().setTitle(problem.getAddress());
        marker.setIcon(getMarkerIcon(problem));
    }

    @Override
    public void onMapReady(GoogleMap map) {
        super.onMapReady(map);

        map.setOnInfoWindowClickListener(this);
        map.setOnInfoWindowCloseListener(this);
    }

    private void setMapTitle(String address) {
        getActivity().setTitle(address);
    }
}

File: app/src/main/java/lt/vilnius/tvarkau/fragments/SingleProblemMapFragment.kt
package lt.vilnius.tvarkau.fragments

import android.os.Bundle
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.Marker
import lt.vilnius.tvarkau.activity.ActivityConstants
import lt.vilnius.tvarkau.entity.Problem
import lt.vilnius.tvarkau.entity.ReportEntity

@Screen(
    navigationMode = NavigationMode.BACK,
    trackingScreenName = ActivityConstants.SCREEN_REPORT_MAP
)
class SingleProblemMapFragment : BaseMapFragment(),
    GoogleMap.OnInfoWindowClickListener,
    GoogleMap.OnInfoWindowCloseListener {

    val problem: Problem
        get() = arguments!!.getParcelable(ARG_REPORT)

    private fun initMapData() {
        placeAndShowMarker(problem)
        setMapTitle(problem.address)
    }

    override fun onInfoWindowClick(marker: Marker) {
        activity!!.onBackPressed()
    }

    override fun onInfoWindowClose(marker: Marker) {
        val problem = marker.tag as Problem

        activity!!.title = problem.address
        marker.setIcon(getMarkerIcon(problem))
    }

    override fun onMapLoaded() {
        super.onMapLoaded()

        googleMap?.setOnInfoWindowClickListener(this)
        googleMap?.setOnInfoWindowCloseListener(this)

        initMapData()
    }

    private fun setMapTitle(address: String?) {
        activity!!.title = address
    }

    companion object {
        private const val ARG_REPORT = "report"

        fun getInstance(reportEntity: ReportEntity): SingleProblemMapFragment {
            return SingleProblemMapFragment().apply {
                arguments = Bundle().apply {
                    putParcelable(ARG_REPORT, reportEntity)
                }
            }
        }
    }
}

